### PR TITLE
Script for updating dapi fallback hash merkle tree values

### LINKED ISSUE #23: Add script to populate current and previous merkle tree data (dAPI fallback tree)

### PR DESCRIPTION
**Pull request by pieterdub**
Closes #23

- [x]  Add script to check if there are new values on `@nodary/utilities` package
- [x]  Update Readme to include instructions to use script

### CODE REVIEW

File: scripts/syncTreeValues.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    process.exit(1);
+  }
+
+  const currentHashData = JSON.parse(fs.readFileSync(currentHashPath, 'utf8'));
+  const currentHashValues = currentHashData.merkleTreeValues ? currentHashData.merkleTreeValues.values : [];
+
+  const values = getExistingTreeValues(merkleTreeName);
+
+  if (isEqual(currentHashValues, values)) {
+    console.info('Current hash file is up to date!');
```
Comment (mcoetzee) [inline scripts/syncTreeValues.js:54]: Very minor: Don't think we need to add exclamation marks 🙂

File: scripts/syncTreeValues.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+const merkleType = process.argv[2];
+
+const MERKLE_TREE_MAPPING = {
+  price: 'dapi-pricing-merkle-tree-root',
+  'dapi management': 'dapi-management-merkle-tree-root',
+  'dapi fallback': 'dapi-fallback-merkle-tree-root',
+  'api integration': 'signed-api-url-merkle-tree-root',
+};
+
+function getExistingTreeValues(merkleTreeName) {
```
Comment (mcoetzee) [inline scripts/syncTreeValues.js:17]: Following prev feedback, there is no need for this function, we can just call `getNodaryFallbackValues()`

File: scripts/syncTreeValues.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+    const updatedData = { ...currentHashData, merkleTreeValues: { values } };
+    fs.writeFileSync(currentHashPath, JSON.stringify(updatedData, null, 4));
+
+    exec('yarn prettier');
+    console.info('Current hash file updated!');
+  }
+}
+
+//#region DAPI Fallback values fetch
```
Comment (mcoetzee) [inline scripts/syncTreeValues.js:71]: Minor: These `#region ` and `#endregion` comments seems a bit unusual?

File: package.json [hunk 4]
```diff
@@ -24,6 +24,7 @@
     "test:gas": "REPORT_GAS=TRUE hardhat test",
     "write-env-example": "node scripts/write-env-example.js",
     "sign": "node scripts/signMerkleTree.js",
+    "sync-tree-values": "node scripts/syncTreeValues.js",
```
Comment (mcoetzee) [inline package.json:27]: Following prev feedback, we can name this less broadly. Perhaps:
```json
"sync-dapi-fallback-values": "node scripts/sync-dapi-fallback-values.js",
```
Also, lets stick to `kebab-case` when naming files.

File: scripts/README.md [hunk 5]
```diff
@@ -74,3 +74,20 @@ The following code snippet shows how we derive the type of Signed API URL Merkle
 ```js
 keccak256(abi.encodePacked('Signed API URL Merkle tree root'));
 ```
+
+## Usage Of Update DApi Fallback hash
```
Comment (mcoetzee) [inline scripts/README.md:78]: This heading isn't too clear to me. What about calling the section: _How to update the dAPI Fallback Merkle Tree_

Step one would be to update the `@nodary/utilities` package.

Also, following prev feedback, remember to update the below

File: scripts/README.md [hunk 6]
```diff
@@ -74,3 +74,20 @@ The following code snippet shows how we derive the type of Signed API URL Merkle
...
+```
+
+For instance, to update the current-hash fo `dAPI fallback Merkle tree` Merkle tree, you would use:
+
+```bash
+yarn sync-tree-values "dapi fallback"
+```
+
+This command checks the values in `current-hash.json` and compares it with what `@nodary/utilities` has. If they are not equal
+It saves the `current-hash.json` to `previous-hash.json` and updates `current-hash.json` with the new values.
```
Comment (mcoetzee) [inline scripts/README.md:93]: Minor: `It` shouldn't be capitalised

**Review by mcoetzee — CHANGES_REQUESTED** [review]

File: scripts/syncTreeValues.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    console.info('Current hash file not found');
+    process.exit(1);
+  }
+
+  const currentHashData = JSON.parse(fs.readFileSync(currentHashPath, 'utf8'));
+  const currentHashValues = currentHashData.merkleTreeValues ? currentHashData.merkleTreeValues.values : [];
+
+  const values = getExistingTreeValues(merkleTreeName);
+
+  if (isEqual(currentHashValues, values)) {
```
[the remark this answers is not in the document]
Comment (pieterdub) [reply scripts/syncTreeValues.js]: Cool cool thanks 🙏

File: scripts/syncTreeValues.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
+const ethers = require('ethers');
+const fs = require('fs');
+const path = require('path');
+const nodaryUtilities = require('@nodary/utilities');
+const { isEqual } = require('lodash');
+const { exec } = require('child_process');
+
+const merkleType = process.argv[2];
+
+const MERKLE_TREE_MAPPING = {
```
[the remark this answers is not in the document]
Comment (pieterdub) [reply scripts/syncTreeValues.js]: Good to know 👍

File: data/dapi-fallback-merkle-tree-root/current-hash.json [hunk 9]
```diff
@@ -1,11 +1,7 @@
...
-  "timestamp": 1676940000,
-  "hash": "",
-  "signatures": {
-    "0x80efDd3bB15F2108C407049C5575490858800D47": "0xfe2cd37f6922ecad2e6277165caf65b0a1ed8b854e21ceac45c8f34ed60e3a1150ed6957d7f3f00d40b9f63ad3016051d18513dfeb95b81f8ff362a7017440a11c",
-    "0x09dEa15D4E1F82787e181F23788Dc3f4b3d3DAE4": "0x23e0b0c2d46d52bba42eb9b14227c3dd43efd80a67929938106ae83729794b986a3deb17ef83be08f66d725144b1183b718c59097d93bb1c693313677bbb2f571c",
-    "0x433171908d5BAB4ac0E860c64D1333F51A310321": "0x33de3c54a8120ebe04a79cbbb15378a5fc91e0577b84d2ffc4c03126f48468a43d928d6b07e2381e6ed47c205f15dc0dc0ac56b99b0ca3916b862828bd5fb48b1c"
-  },
+  "timestamp": 1699287899069,
+  "hash": "0x508b6d6bd706095091bc3cb96716d12a8e803f3b2b85a7cdc0274486452de6ec",
+  "signatures": {},
```
Comment (mcoetzee) [inline data/dapi-fallback-merkle-tree-root/current-hash.json:4]: We need to keep the current `signatures` and `timestamp`.

**Review by mcoetzee — APPROVED** [review]
Just got one last comment, so I'm approving so long 👍

**Pull request merged**